### PR TITLE
Adding handling for AutocompletePrediction's structured formatting.

### PR DESCRIPTION
**Pull request by domesticmouse**
This is a fix for https://github.com/googlemaps/google-maps-services-java/issues/244

### CODE REVIEW

File: src/main/java/com/google/maps/model/AutocompletePrediction.java [hunk 1]
```diff
@@ -59,9 +59,19 @@
...
   }
 
+  /**
+   * Describe the location of the entered term in the prediction result text, so that the term can
+   * be highlighted if desired.
+   */
   public MatchedSubstring matchedSubstrings[];
 
   /**
+   * Structured formatting contains a description of how the autocomplete query matched the returned
```
Comment (apjanke) [inline src/main/java/com/google/maps/model/AutocompletePrediction.java]: I'd simplify to `A description of how the autocomplete query matched the returned result.`. Leading with "Structured formatting contains..." is basically repeating the class name.

File: src/main/java/com/google/maps/model/AutcompleteStructuredFormatting.java [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+/** Represents the Structured Formatting field of an {@see AutocompletePrediction}. */
+public class AutcompleteStructuredFormatting {
+
+  /** Contains the main text of a prediction, usually the name of the place. */
+  public String mainText;
+
+  /** Contains an array of where the query matched the returned main text. */
+  public AutocompletePrediction.MatchedSubstring mainTextMatchedSubstrings[];
+
+  /** Contains the secondary text of a prediction, usually the location of the place. */
```
Comment (apjanke) [inline src/main/java/com/google/maps/model/AutcompleteStructuredFormatting.java]: I'd omit "Contains".

**Comment by apjanke** [comment]
Thanks for the interest!

I made some minor style comments.

There's maybe a substantive issue to cover here, too: Are the `length` and `offset` values in `MatchedSubstring` in Unicode characters, or Java `char`s? Those are often, but not always, the same thing – Java `char`s are UTF-16 code units, not necessarily full Unicode characters/code points, as characters outside the BMP may be represented by surrogate pairs. Normally in all-Java APIs this isn't an issue, but since we're dealing with an external API here, there may be an "impedance mismatch". I've seen other places in the Google Maps API docs where it describes offsets as being "in Unicode characters". If that's the case here too, it's worth noting. If `MatchedSubstring` is carrying Unicode-character offsets, instead of `char` offsets, that might be a surprise to your average Java developer. May be worth noting in the Javadoc which it is.

(I don't know how likely this difference is to crop up in practice for Google Maps, because I don't know how far its localization reaches, but it might be worth noting for full correctness.)

**Comment by domesticmouse** [comment]
The APIs are served in UTF8 I believe, and the length and offsets should be Unicode characters, but I'm unsure whether we document this anywhere. 

At this point I'm copying across the documentation from the Places API documentation, as that is what is published, and I'm uncomfortable going above and beyond that.

That said, if someone tests this, and finds that we aren't serving length and offset as unicode characters, then please raise a bug on our [issue tracker](https://developers.google.com/maps/support/#issue_tracker). 

This client library is merely passing through what the APIs return, if the APIs are returning data that is incorrect then we have to fix the APIs themselves, not this client library.

File: src/main/java/com/google/maps/model/AutcompleteStructuredFormatting.java [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+ * Unless required by applicable law or agreed to in writing, software distributed under
+ * the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF
+ * ANY KIND, either express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+package com.google.maps.model;
+
+/** Represents the Structured Formatting field of an {@see AutocompletePrediction}. */
+public class AutcompleteStructuredFormatting {
```
Comment (domesticmouse) [inline src/main/java/com/google/maps/model/AutcompleteStructuredFormatting.java]: Done, and yes, that's quite a typo right there...

File: src/main/java/com/google/maps/model/AutcompleteStructuredFormatting.java [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+
+  /** Contains the main text of a prediction, usually the name of the place. */
```
Comment (domesticmouse) [inline src/main/java/com/google/maps/model/AutcompleteStructuredFormatting.java]: Done

File: src/main/java/com/google/maps/model/AutcompleteStructuredFormatting.java [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+  public String mainText;
+
+  /** Contains an array of where the query matched the returned main text. */
```
Comment (domesticmouse) [inline src/main/java/com/google/maps/model/AutcompleteStructuredFormatting.java]: done

File: src/main/java/com/google/maps/model/AutcompleteStructuredFormatting.java (hunk 2 above)
Comment (domesticmouse) [inline src/main/java/com/google/maps/model/AutcompleteStructuredFormatting.java]: done

File: src/main/java/com/google/maps/model/AutocompletePrediction.java [hunk 6]
```diff
@@ -59,9 +59,19 @@
     public int offset;
   }
 
+  /**
+   * Describe the location of the entered term in the prediction result text, so that the term can
```
Comment (domesticmouse) [inline src/main/java/com/google/maps/model/AutocompletePrediction.java]: done

File: src/main/java/com/google/maps/model/AutocompletePrediction.java (hunk 1 above)
Comment (domesticmouse) [inline src/main/java/com/google/maps/model/AutocompletePrediction.java]: done

**Comment by apjanke** [comment]
Oh, I don't think it's an issue of correctness with the APIs, just a matter of setting expectations for Java-centric developers who use this library. From what I've seen, most Java APIs seem to use Java `char` counts for string offsets, and mostly overlook the fact that Java `char`s are not exactly equivalent to Unicode characters. That is, coders expect offsets and lengths to index in to `String`s using the `charAt()` method or array indexing into the corresponding `char[]` array, and may need to be given notice that they should be using the `offsetByCodePoints()` method instead. I think simply noting in the Javadoc for those fields that they are offsets in Unicode characters (as opposed to Java `char`s) would be sufficient and appropriate. Like the existing documentation for `AutocompletePrediction.terms` does [here](https://github.com/googlemaps/google-maps-services-java/blob/1616f4f2c16c24dc4db844945f4905133760514f/src/main/java/com/google/maps/model/AutocompletePrediction.java#L45); I'd just add "`...measured in Unicode characters` to the new length/offset fields' Javadoc.

Sorry if I'm belaboring a point that's already been gone over here in the past.

**Comment by domesticmouse** [comment]
Valid commentary, I'm not totally up to speed on I18N concerns, so I value this feedback. And I'd totally forgotten that Java is UTF16. I've obviously spent too long with Go and it's Unicode runes....

**Comment by domesticmouse** [comment]
How's that look @apjanke?

**Comment by apjanke** [comment]
Yep, LGTM!

**Comment by domesticmouse** [comment]
Thanks for your careful reviews @apjanke!

**Pull request merged**